[WIP] Add UniDiffuser model and pipeline (#2963)
huggingface/diffusers: docs/source/en/api/pipelines/unidiffuser.mdx

Resources:

* [Paper](https://arxiv.org/abs/2303.06555).
* [Original Code](https://github.com/thu-ml/unidiffuser).

Available Checkpoints are:
- *UniDiffuser-v0 (512x512 resolution)* [thu-ml/unidiffuser-v0](https://huggingface.co/thu-ml/unidiffuser-v0)
- *UniDiffuser-v1 (512x512 resolution)* [thu-ml/unidiffuser-v1](https://huggingface.co/thu-ml/unidiffuser-v1)

This pipeline was contributed by our community member [dg845](https://github.com/dg845).

## Available Pipelines:

> | Pipeline | Tasks | Demo
> |---|---|:---:|
> | [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* | [🤗 Spaces](https://huggingface.co/spaces/thu-ml/unidiffuser) |

Reviewer: ```suggestion
| Pipeline | Tasks | Demo | Colab |
|:---:|:---:|:---:|:---:|
| [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*,<br> *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* | [🤗 Spaces](https://huggingface.co/spaces/thu-ml/unidiffuser) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/unidiffuser.ipynb) |
```

Resources:

* [Paper](https://arxiv.org/abs/2303.06555).
* [Original Code](https://github.com/thu-ml/unidiffuser).

Available Checkpoints are:
- *UniDiffuser-v0 (512x512 resolution)* [dg845/unidiffuser-diffusers-v0](https://huggingface.co/dg845/unidiffuser-diffusers-v0)
- *UniDiffuser-v1 (512x512 resolution)* [dg845/unidiffuser-diffusers-v1](https://huggingface.co/dg845/unidiffuser-diffusers)

## Available Pipelines:

| Pipeline | Tasks | Demo
|---|---|:---:|
| [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* |  |

## Usage Examples

Because the UniDiffuser model is trained to model the joint distribution of (image, text) pairs, it is capable of performing a diverse range of generation tasks.

### Unconditional Image and Text Generation

Unconditional generation (where we start from only latents sampled from a standard Gaussian prior) from a `UniDiffuserPipeline` will produce a (image, text) pair:

```python
import torch

from diffusers import UniDiffuserPipeline

device = "cuda"

> model_id_or_path = "dg845/unidiffuser-diffusers"

Reviewer: To change. Internally discussing to transfer the weights.

print(text)
```

This is also called "joint" generation in the UniDiffusers paper, since we are sampling from the joint image-text distribution.

Note that the generation task is inferred from the inputs used when calling the pipeline.
It is also possible to manually specify the unconditional generation task ("mode") manually with [`UniDiffuserPipeline.set_joint_mode`]:

```python
# Equivalent to the above.
pipe.set_joint_mode()
sample = pipe(num_inference_steps=20, guidance_scale=8.0)
```

When the mode is set manually, subsequent calls to the pipeline will use the set mode without attempting the infer the mode.
You can reset the mode with [`UniDiffuserPipeline.reset_mode`], after which the pipeline will once again infer the mode.

You can also generate only an image or only text (which the UniDiffuser paper calls "marginal" generation since we sample from the marginal distribution of images and text, respectively):

```python
# Unlike other generation tasks, image-only and text-only generation don't use classifier-free guidance
# Image-only generation
pipe.set_image_mode()
sample_image = pipe(num_inference_steps=20).images[0]
# Text-only generation
pipe.set_text_mode()

> sample_text = pipe(num_inference_steps=20).text[0]

Reviewer: Should we advise the users that text generation is currently buggy? Or have we fixed it already?

Author: The image-to-text generation issue has been fixed, so all generation tasks should be working as expected.

t2i_image = sample.images[0]
t2i_image.save("unidiffuser_text2img_sample_image.png")
```

The `text2img` mode requires that either an input `prompt` or `prompt_embeds` be supplied. You can set the `text2img` mode manually with [`UniDiffuserPipeline.set_text_to_image_mode`].

### Image-to-Text Generation

Similarly, UniDiffuser can also produce text samples given an image (image-to-text or image-conditioned text generation):

```python
import requests
import torch
from PIL import Image
from io import BytesIO

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "dg845/unidiffuser-diffusers"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Image-to-text generation
image_url = "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/unidiffuser/unidiffuser_example_image.jpg"
response = requests.get(image_url)
init_image = Image.open(BytesIO(response.content)).convert("RGB")
init_image = init_image.resize((512, 512))

sample = pipe(image=init_image, num_inference_steps=20, guidance_scale=8.0)
i2t_text = sample.text[0]

> print(text)

Reviewer: Is it currently working as expected? I am sorry if I missed any of your messages around its fix. If not, I think it might make sense to add a little disclaimer here.

Author: The image-to-text generation issue has been fixed, so all generation tasks should be working as expected.

sample = pipe(image=init_image, num_inference_steps=20, guidance_scale=8.0)
i2t_text = sample.text[0]
print(text)

# 2. Text-to-image generation
sample = pipe(prompt=i2t_text, num_inference_steps=20, guidance_scale=8.0)
final_image = sample.images[0]
final_image.save("unidiffuser_image_variation_sample.png")
```

### Text Variation

Similarly, text variation can be performed on an input prompt with a text-to-image generation followed by a image-to-text generation:

```python
import torch

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "dg845/unidiffuser-diffusers"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Text variation can be performed with a text-to-image generation followed by a image-to-text generation:
# 1. Text-to-image generation
prompt = "an elephant under the sea"

sample = pipe(prompt=prompt, num_inference_steps=20, guidance_scale=8.0)
t2i_image = sample.images[0]
t2i_image.save("unidiffuser_text2img_sample_image.png")

# 2. Image-to-text generation
sample = pipe(image=t2i_image, num_inference_steps=20, guidance_scale=8.0)
final_prompt = sample.text[0]
print(final_prompt)

> ```

Reviewer: Separate examples look super!

Resources:

* [Paper](https://arxiv.org/abs/2303.06555).
* [Original Code](https://github.com/thu-ml/unidiffuser).

Available Checkpoints are:
- *UniDiffuser-v0 (512x512 resolution)* [thu-ml/unidiffuser-v0](https://huggingface.co/thu-ml/unidiffuser-v0)
- *UniDiffuser-v1 (512x512 resolution)* [thu-ml/unidiffuser-v1](https://huggingface.co/thu-ml/unidiffuser-v1)

This pipeline was contributed by our community member [dg845](https://github.com/dg845).

## Available Pipelines:

| Pipeline | Tasks | Demo
|---|---|:---:|

> | [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* |  |

Reviewer: ```suggestion
| Pipeline | Tasks | Demo
|---|---|:---:|
| [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* | [🤗 Spaces](https://huggingface.co/spaces/thu-ml/unidiffuser) |

For now, let's add a link to the original demo. @hysts is working on to change the demo to have `diffusers` usage.

Reviewer: I am also preparing a Colab Notebook. Will share here soon :)

## Usage Examples

Because the UniDiffuser model is trained to model the joint distribution of (image, text) pairs, it is capable of performing a diverse range of generation tasks.

### Unconditional Image and Text Generation

> Unconditional generation (where we start from only latents sampled from a standard Gaussian prior) from a `UniDiffuserPipeline` will produce a (image, text) pair:

Reviewer: ```suggestion
Unconditional generation (where we start from only latents sampled from a standard Gaussian prior) from a [`UniDiffuserPipeline`] will produce a (image, text) pair:
```

So, that the hyperlink is automatically rendered.

```

### Text-to-Image Generation

UniDiffuser is also capable of sampling from conditional distributions; that is, the distribution of images conditioned on a text prompt or the distribution of texts conditioned on an image.
Here is an example of sampling from the conditional image distribution (text-to-image generation or text-conditioned image generation):

```python
import torch

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "thu-ml/unidiffuser-v1"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Text-to-image generation
prompt = "an elephant under the sea"

sample = pipe(prompt=prompt, num_inference_steps=20, guidance_scale=8.0)
t2i_image = sample.images[0]
t2i_image.save("unidiffuser_text2img_sample_image.png")
```

The `text2img` mode requires that either an input `prompt` or `prompt_embeds` be supplied. You can set the `text2img` mode manually with [`UniDiffuserPipeline.set_text_to_image_mode`].

### Image-to-Text Generation

Similarly, UniDiffuser can also produce text samples given an image (image-to-text or image-conditioned text generation):

```python

> import requests
> import torch
> from PIL import Image
> from io import BytesIO
> 
> from diffusers import UniDiffuserPipeline
> 
> device = "cuda"
> model_id_or_path = "thu-ml/unidiffuser-v1"
> pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
> pipe.to(device)
> 
> # Image-to-text generation
> image_url = "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/unidiffuser/unidiffuser_example_image.jpg"
> response = requests.get(image_url)
> init_image = Image.open(BytesIO(response.content)).convert("RGB")
> init_image = init_image.resize((512, 512))
> 
> sample = pipe(image=init_image, num_inference_steps=20, guidance_scale=8.0)
> i2t_text = sample.text[0]
> print(text)

Reviewer: ```suggestion
from diffusers import UniDiffuserPipeline
from diffusers.utils import load_image

device = "cuda"
model_id_or_path = "thu-ml/unidiffuser-v1"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Image-to-text generation
image_url = "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/unidiffuser/unidiffuser_example_image.jpg"
init_image = load_image(image_url).resize((512, 512))
sample = pipe(image=init_image, num_inference_steps=20, guidance_scale=8.0)
i2t_text = sample.text[0]
print(text)
```

Reduces the LoC :)

```

> The `img2text` mode requires that an input `image` be supplied. You can set the `img2text` mode manually with [`UniDiffuser.set_image_to_text_mode`].

Reviewer: ```suggestion
The `img2text` mode requires that an input `image` be supplied. You can set the `img2text` mode manually with [`UniDiffuserPipeline.set_image_to_text_mode`].
```

### Image Variation

The UniDiffuser authors suggest performing image variation through a "round-trip" generation method, where given an input image, we first perform an image-to-text generation, and the perform a text-to-image generation on the outputs of the first generation.
This produces a new image which is semantically similar to the input image:

```python
import requests
import torch
from PIL import Image
from io import BytesIO

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "thu-ml/unidiffuser-v1"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Image variation can be performed with a image-to-text generation followed by a text-to-image generation:
# 1. Image-to-text generation
image_url = "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/unidiffuser/unidiffuser_example_image.jpg"
response = requests.get(image_url)
init_image = Image.open(BytesIO(response.content)).convert("RGB")

> init_image = init_image.resize((512, 512))

Reviewer: I guess we can follow the same as https://github.com/huggingface/diffusers/pull/2963/files#r1205061596 for loading and resizing the image?

sample = pipe(image=init_image, num_inference_steps=20, guidance_scale=8.0)
i2t_text = sample.text[0]

> print(text)

Reviewer: Nit: should be `i2t_text`.

# 2. Text-to-image generation
sample = pipe(prompt=i2t_text, num_inference_steps=20, guidance_scale=8.0)
final_image = sample.images[0]
final_image.save("unidiffuser_image_variation_sample.png")
```

### Text Variation

Similarly, text variation can be performed on an input prompt with a text-to-image generation followed by a image-to-text generation:

```python
import torch

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "thu-ml/unidiffuser-v1"
pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
pipe.to(device)

# Text variation can be performed with a text-to-image generation followed by a image-to-text generation:
# 1. Text-to-image generation
prompt = "an elephant under the sea"

sample = pipe(prompt=prompt, num_inference_steps=20, guidance_scale=8.0)
t2i_image = sample.images[0]
t2i_image.save("unidiffuser_text2img_sample_image.png")

# 2. Image-to-text generation
sample = pipe(image=t2i_image, num_inference_steps=20, guidance_scale=8.0)
final_prompt = sample.text[0]
print(final_prompt)
```

## UniDiffuserPipeline
[[autodoc]] UniDiffuserPipeline
	- all
	- __call__

> ## ImageTextPipelineOutput

Reviewer: This should go here: 

https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/outputs.mdx

Reviewer: Also wanted to know if there's any argument control the number of images / text I wanted to generate as a part of the variation mode.

Reviewer: I can control the `num_images_per_prompt` in the text-to-image mode, so that's settled. But what about text variation?

Author: For modes which generate only text (`img2text` and `text`), there's an analogous `num_prompts_per_image` argument to `__call__ `. So when you perform the second `img2text` generation for text variation you can specify `num_prompts_per_image > 1` to get multiple text variation samples.

Reviewer: Thanks!

Author: > This should go here:
> 
> https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/outputs.mdx

It feels more natural to me to have the documentation for `ImageTextPipelineOutput` alongside `ImagePipelineOutput`, which is at the [`Diffusion Pipeline`](https://huggingface.co/docs/diffusers/main/en/api/diffusion_pipeline#diffusers.ImagePipelineOutput) doc page.

Author: I've gone ahead and moved the `ImageTextPipelineOutput` documentation to `/api/diffusion_pipeline.mdx` (alongside the `ImagePipelineOutput` and `AudioPipelineOutput` documentation). Let me know if it would be better somewhere else (for example, at `/api/outputs.mdx` as originally suggested) :).

Reviewer: I understand. But we will soon update that too :)

Author: I see, so should I move it to `/api/outputs`? Or is it fine to leave it at `/api/diffusion_pipeline` for now?

Resources:

* [Paper](https://arxiv.org/abs/2303.06555).
* [Original Code](https://github.com/thu-ml/unidiffuser).

Available Checkpoints are:
- *UniDiffuser-v0 (512x512 resolution)* [dg845/unidiffuser-diffusers-v0](https://huggingface.co/dg845/unidiffuser-diffusers-v0)
- *UniDiffuser-v1 (512x512 resolution)* [dg845/unidiffuser-diffusers-v1](https://huggingface.co/dg845/unidiffuser-diffusers)

## Available Pipelines:

| Pipeline | Tasks | Demo
|---|---|:---:|
| [UniDiffuserPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_unidiffuser.py) | *Joint Image-Text Gen*, *Text-to-Image*, *Image-to-Text*, *Image Gen*, *Text Gen*, *Image Variation*, *Text Variation* |  |

## Usage Example

```python
import requests
import torch
from PIL import Image
from io import BytesIO

from diffusers import UniDiffuserPipeline

device = "cuda"
model_id_or_path = "dg845/unidiffuser-diffusers"

> pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path)

Reviewer: For my understanding, does this currently support (as in terms of quality):

```python
import torch

pipe = UniDiffuserPipeline.from_pretrained(model_id_or_path, torch_dtype=torch.float16)
```

This is almost a defacto standard for performing inference in `diffusers`.

Author: Currently there are no errors with mixed precision inference, but there may be some issues with quality (which may or may not be related to mixed precision); see discussion about generation quality.

Reviewer: Hmm. Interesting.

Does using mixed-precision in the original implementation also suffer from this issue? Also, does it happen for both the checkpoints? 

Maybe there needs to be intermediate upcasting (softmax ops, for example) to prevent numerical overflows? For pipelines like [`StableUnCLIPPipeline`](https://huggingface.co/docs/diffusers/api/pipelines/stable_unclip) we use this. See: 

https://huggingface.co/stabilityai/stable-diffusion-2-1-unclip-small/blob/main/unet/config.json

Author: TL;DR: image-to-text samples may be messed up (other modes are probably fine?), which is true for both the v0 and v1 checkpoints. I think this is separate from numerical stability issues but haven't been able to test the full checkpoint in full precision. Using something like `upcast_attention=True` doesn't seem to solve the issue.

----

To be more precise about the issue, when doing image-to-text generation, the output text seems to be messed up. In particular, for the `diffusers` version of the UniDiffuser-v1 model, if I generate samples (with `num_inference_steps=20` and `guidance_scale=8.0`) in the different modes that produce text samples:

```bash
# Unconditional ("joint") image and text generation
A bowl of pasta with tomatoes and greens
```
![unidiffuser_sample_joint_image](https://github.com/huggingface/diffusers/assets/58458699/9f581b22-00ef-4c39-9909-c06ca49aa944)

Both the generated text and image seem reasonable, and they also match each other.

Similarly, in text-only generation mode, we get

```bash
# Text-only ("marginal text") generation
A path lined with blue pines in the woods
```

But in image-to-text mode, when I use the [UniDiffuser example image](https://github.com/thu-ml/unidiffuser/blob/main/assets/space.jpg) as the input image, I get the following output:

```bash
# Image-to-text generation
...: 'Inclothes: 'Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes: Inclothes
```

I've generated multiple samples (all in mixed precision) and they all end up like the above sample. Increasing `num_inference_steps` neither makes the text samples more fluent nor more accurate in describing the image.

Unfortunately, since I'm not able to generate samples from the full checkpoint using the original implementation on my computer, I'm not sure whether this is normal for the model or not. For example, when we ran the original implementation to generate expected output for the image-to-text slow test (which uses the same image and `num_inference_steps=3`), the text output is

```
T CL CL CL TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK TICK T
```

so these types of outputs might not be unusual. On the other hand, the original model definitely generates good samples at least sometimes (e.g. the image-to-text samples from the [original paper](https://arxiv.org/pdf/2303.06555.pdf)).

I think this is distinct from the numerical stability issue mentioned in a previous comment (https://github.com/huggingface/diffusers/pull/2963#issuecomment-1547028401), although presumably the two issues interact with each other. I've tested a little with `upcast_attention=True` and it doesn't seem to solve the image-to-text bad samples issue. (I haven't tested whether it will solve the numerical issues in `torch 1.x` yet.)

> Does using mixed-precision in the original implementation also suffer from this issue?

My understanding is that the original implementation loads the models in full precision but then [runs the sampling loop in half precision](https://github.com/thu-ml/unidiffuser/blob/main/sample_multi_v1.py#L302). I haven't tested modifying this behavior yet. As for whether it has the image-to-text sample issue, see discussion above.

> Also, does it happen for both the checkpoints?

The image-to-text sample issue happens for both the UniDiffuser-v0 and UniDiffuser-v1 checkpoints.

Reviewer: Thanks for your detailed answer. If image-to-text generation issue happens with the original implementation code as well, do you think it could make sense to check with the authors? 

Also, for testing purposes, I have gone ahead and added you to this Space (that access to an A10G): https://huggingface.co/spaces/unidiffuser-testing/unidiffuser-testing.

Hopefully, that will help you to run the tests on the full checkpoint?

Author: I think I have debugged the error, turns out when encoding the image using the VAE I forgot to scale the latents by the VAE's scaling factor 😅.

Reviewer: Damn it that little bugger! Can be very brutal when missed!

pipe.to(device)

> # Joint image-text generation. The generation task is automatically inferred.

Reviewer: Worth elaborating what does this mean. 

Does this mean it's unconditional generation?

Reviewer: Yup certainly.

Reviewer: Maybe to be ultra clear - "Unconditional image and text generation"

If the text is generated based on the generated images, the "Unconditional image-text generation" makes more sense.

sample = pipe(num_inference_steps=20, guidance_scale=8.0)
image = sample.images[0]
text = sample.text[0]
image.save("unidiffuser_sample_joint_image.png")
print(text)

# The mode can be set manually. The following is equivalent to the above:
pipe.set_joint_mode()
sample2 = pipe(num_inference_steps=20, guidance_scale=8.0)

# Note that if you set the mode manually the pipeline will no longer attempt
# to automatically infer the mode. You can re-enable this with reset_mode().

> pipe.reset_mode()

Reviewer: +300 to split like that. It becomes so much easier to follow that way for a user.
